File: client/java/src/avt/Message.java
package avt;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public final class Message {
   public byte command;
   private ByteArrayOutputStream b = null;
   private DataOutputStream dos = null;
   private ByteArrayInputStream d = null;
   private DataInputStream dis = null;

   public Message() {
   }

   public Message(byte var1) {
      this.command = var1;
      this.b = new ByteArrayOutputStream();
      this.dos = new DataOutputStream(this.b);
   }

   public Message(byte var1, byte[] var2) {
      this.command = var1;
      this.d = new ByteArrayInputStream(var2);
      this.dis = new DataInputStream(this.d);
   }

   public final byte[] getData() {
      return this.b.toByteArray();
   }

   public final DataInputStream reader() {
      return this.dis;
   }

   public final DataOutputStream writer() {
      return this.dos;
   }

   public final void cleanup() {
      try {
         if (this.dis != null) {
            this.dis.close();
         }

         if (this.dos != null) {
            this.dos.close();
            return;
         }
      } catch (IOException var2) {
      }

   }
}

File: server-node/src/net/Message.js
/** Port của avatar/network/Message.java */
import { DataOutputStream, DataInputStream } from './JavaIO.js';

export class Message {
  /**
   * @param {number} command
   * @param {Buffer=} data nếu có -> message để ĐỌC; không có -> message để GHI
   */
  constructor(command, data = null) {
    this.command = ((Number(command) | 0) << 24) >> 24; // ép về byte có dấu
    if (data != null) {
      this.is = new DataInputStream(data);
      this.os = null;
    } else {
      this.os = new DataOutputStream();
      this.is = null;
    }
  }

  getCommand() {
    return this.command;
  }

  setCommand(cmd) {
    this.command = ((Number(cmd) | 0) << 24) >> 24;
  }

  getData() {
    return this.os ? this.os.toBuffer() : Buffer.alloc(0);
  }

  /** @returns {DataInputStream} */
  reader() {
    return this.is;
  }

  /** @returns {DataOutputStream} */
  writer() {
    return this.os;
  }

  cleanup() {
    // Không cần giải phóng thủ công như Java, giữ API cho tương thích
  }
}

export default Message;
